lib/framework/output.ts

```diff
@@ -0,0 +1,160 @@
+ */
+export function outputMessage(msg: string) {
+	// Messages go to STDERR
+	console.error(msg);
+}
+
+/**
+ * Output result data to STDOUT
+ *  Supports:
+ *   - arrays of items (displayed in a tabular way),
+ *   - single items (displayed in a field per row format).
+ *
+ * @param data Array of data objects to output
+ * @param fields Array of fieldnames, specifying the fields and display order
+ * @param options Output options
+ */
+export async function outputData(
+	data: any[] | {},
+	fields: string[],
+	options: DataOutputOptions | DataSetOutputOptions,
+) {
+	if (Array.isArray(data)) {
+		await outputDataSet(data, fields, options as DataSetOutputOptions);
+	} else {
+		await outputDataItem(data, fields, options as DataOutputOptions);
+	}
+}
+
+/**
+ * Wraps the cli.ux table implementation, to output tabular data
+ *
+ * @param data Array of data objects to output
+ * @param fields Array of fieldnames, specifying the fields and display order
+ * @param options Output options
+ */
+async function outputDataSet(
+	data: any[],
+	fields: string[],
+	options: DataSetOutputOptions,
+) {
+	// Oclif expects fields to be specified in the format used in table headers (though lowercase)
+	// By replacing underscores with spaces here, we can support both header format and actual field name
+	// (e.g. as seen in json output).
+	options.fields = options.fields?.replace(/_/g, ' ');
+	options.filter = options.filter?.replace(/_/g, ' ');
+	options.sort = options.sort?.replace(/_/g, ' ');
+
+	getCliUx().table(
+		data,
+		// Convert fields array to column object keys
+		// that cli.ux expects.  We can later add support
+		// for both formats if beneficial
+		fields.reduce((ac, a) => ({ ...ac, [a]: {} }), {}),
+		{
+			...options,
+			...(options.json
+				? {
+						output: 'json',
+				  }
+				: {}),
+			columns: options.fields,
+			extended: options.verbose,
+			printLine,
+		},
+	);
+}
+
+/**
+ * Outputs a single data object (like `resin-cli-visuals table.vertical`),
+ *  but supporting a subset of options from `cli-ux table` (--json and --fields)
+ *
+ * @param data Array of data objects to output
+ * @param fields Array of fieldnames, specifying the fields and display order
+ * @param options Output options
+ */
+async function outputDataItem(
+	data: any,
+	fields: string[],
+	options: DataOutputOptions,
+) {
+	const outData: typeof data = {};
+
+	// Convert comma separated list of fields in `options.fields` to array of correct format.
+	// Note, user may have specified the true field name (e.g. `some_field`),
+	// or the format displayed in headers (e.g. `Some field`, case insensitive).
+	const userSelectedFields = options.fields?.split(',').map((f) => {
+		return f.toLowerCase().trim().replace(/ /g, '_');
+	});
+
+	// Order and filter the fields based on `fields` parameter and `options.fields`
+	fields.forEach((fieldName) => {
+		if (!userSelectedFields || userSelectedFields.includes(fieldName)) {
+			outData[fieldName] = data[fieldName];
```

Comment:
I had to check this, as I thought it would diverge from oclif behavior.  I thought that the `--fields` flag only specified the fields to be displayed, but that the order was still dictated by the `fields` param.  In that case this change would have altered that.  However it turns out that oclif does indeed take the order from the `--fields` flag, so my implementation is wrong.

One way it does diverge from oclif though is if the user specifies a non existent field.  In oclif (and my original implementation) non existent fields are ignored.  However, in your version it would be added to the output data with a value of `undefined`.  I'll modify your concise version to change that.